### PR TITLE
Add ability for most specific header mutations to take precedence

### LINKED ISSUE #8450: allow reversing the order of the evaluateHeaders() calls
Currently, adding or removing request/response headers is hard-coded to the following order [0]:

* route specific
* vhost specific
* connection manager level

This means that you can't set global configs (e.g.: vhost or conn manager level) and then unset them at the route specific level.

Would it be acceptable to have a  config option -- say `reverse_header_evaluation_order` -- to support the reverse order (from least to most specific: conn manager -> vhost -> route)?

Or would this break anything else?

cc: @mattklein123 

[0] https://github.com/envoyproxy/envoy/blob/master/source/common/router/config_impl.cc#L386

### PR DESCRIPTION
**Pull request by rgs1**
This allows changing the precedence of global/vhost/route level
configs when adding/removing headers.

Fixes #8450.

Signed-off-by: Raul Gutierrez Segales <rgs@pinterest.com>

### CODE REVIEW

File: include/envoy/router/router.h [hunk 1]
```diff
@@ -430,6 +430,11 @@ class VirtualHost {
    * @return bool whether to include the request count header in upstream requests.
    */
   virtual bool includeAttemptCount() const PURE;
+
+  /**
+   * @return bool whether to reverse the header evaluation order.
```
[the remark this answers is not in the document]
Comment (rgs1) [reply include/envoy/router/router.h]: yeah good point -- i'll steal some of the wording from protobuf property description:

```
  // By default, the headers that should be added/removed are evaluated in the following order:
  //
  // * route-level headers
  // * virtual host level headers
  // * and finally global connection manager level headers
  //
  // To allow setting overrides at the route or virtual host level, this order can be reversed
  // by setting this option to true. Defaults to false.
```

**Comment by rgs1** [comment]
> At the API this LGTM assuming we don't think there will ever be a need for custom order?

Deferring to @htuch  for that one... (as said, for our use-cases reversing is good enough).

Another API question is how do we feel about setting this at the vhost level? I think it's the right level because:

a) global is to coarse 
b) route is to granular and repetitive 

thoughts?

**Comment by rgs1** [comment]
Not ready for review yet; there's lingering references to "reverse order" that should be renamed to "eval most specific first".

**Comment by rgs1** [comment]
@htuch @mattklein123 i think this is ready for another pass, PTAL.

File: api/envoy/api/v2/rds.proto [hunk 2]
```diff
@@ -108,6 +108,18 @@ message RouteConfiguration {
...
+  // By default, headers that should be added/removed are evaluated from most to least specific:
+  //
+  // * route level
+  // * virtual host level
+  // * connection manager level
+  //
+  // To allow setting overrides at the route or virtual host level, this order can be reversed
+  // by setting this option to true. Defaults to false.
+  //
+  // Note: in the v3 API, this will default to true.
```
Comment (mattklein123) [inline api/envoy/api/v2/rds.proto]: Can you make this a `next-major-version` comment tag (see elsewhere).

Also, if we are going to switch the default should this use `BoolValue`?

Comment (rgs1) [reply api/envoy/api/v2/rds.proto]: Done.

**Comment by rgs1** [comment]
Hmm:

```
@envoy//test/integration:protocol_integration_test                       FAILED in 3 out of 3 in 80.3s
  Stats over 3 runs: max = 80.3s, min = 77.0s, avg = 78.2s, dev = 1.5s
```

**Pull request merged**